# Welcome to OpenOmics's documentation!

.. image:: https://github.com/BioMeCIS-Lab/OpenOmics/raw/master/openomics_web/assets/openomics_logo.png

.. image:: https://img.shields.io/pypi/v/OpenOmics.svg
:target: https://pypi.python.org/pypi/OpenOmics

.. image:: https://readthedocs.org/projects/openomics/badge/?version=latest
:target: https://openomics.readthedocs.io/en/latest/?badge=latest
:alt: Documentation Status

.. image:: https://github.com/BioMeCIS-Lab/OpenOmics/actions/workflows/python-package.yml/badge.svg?branch=master
:target: https://github.com/BioMeCIS-Lab/OpenOmics/actions/workflows/python-package.yml

.. image:: https://codecov.io/gh/BioMeCIS-Lab/OpenOmics/branch/master/graph/badge.svg
:target: https://codecov.io/gh/BioMeCIS-Lab/OpenOmics

This Python package provide a series of tools to integrate and query the genomics, transcriptomics, proteomics, and
clinical data (aka, the multi-omics data). With scalable data-frame manipulation tools, OpenOmics facilitates the common
coding tasks when preparing data for RNA-seq bioinformatics analysis.

- Free software: MIT license
- Documentation: https://openomics.readthedocs.io.

## Features

- Provides a bioinformatics workflow to generate integrative results from multi-omics data.
- Facilitates integration of various bio-databases, multi-omics expression, genomics, and clinical data.
- Highly flexible to different data types and missing data.
- Provides researchers with means to consistently store and explore their experimental datasets.
- Enables scalable performance with parallel computing, while easily configurable to deploy on both single machine and a
  cluster.

## Table of Content

```{toctree}
---
maxdepth: 2
---
installation
   usage
   modules
   contributing
```

## Indices and tables

* :ref:`genindex`
* :ref:`modindex`
* :ref:`search`
